### ISSUE TITLE
GraphQL Subscription does not close sockets to Load Balancer

### ISSUE THREAD

**Issue by STRATZ-Ken** [graphql-dotnet/server#621]
Hello Group, 

Great work with V5.0.  Lots of great improvements.  

Having an issue with the latest WebSockets (Subscriptions).  Prior to V5, #255 was an on-going issue, where we would get these constant disconnects in an aborted status.  This is now gone. However, I have noticed that when behind a load balancer we have some major issues with sockets not being closed on the .Net side.  

![image](https://user-images.githubusercontent.com/24981405/134602812-b4ea24d6-0564-4e5c-807c-9e5eecca7596.png)

This is my server running for just a few minutes, as you can see 4,600~ connections are open.  We currently have about 40 people on the site, thus, its not possible.  How did it get so high? Watch. Now, what I will do is force my load balancer to crash, if we did have 4,600 connections, and if .net releases the connections, we would see it drop to ~0.

![image](https://user-images.githubusercontent.com/24981405/134602936-8af83622-d297-453d-9270-97713bd17ab9.png)

Here is what we get, the existing connections stay open, and an additional ~300 connections are created. Meaning that the connection is not being closed on the .Net side when there is no way the connection can still be open as I crashed the load balancer.  

I can see my DevTools is working correctly to the site (React) is re-creating the connection after the failure of a socket (Reboot).

![image](https://user-images.githubusercontent.com/24981405/134603061-4b9e0467-3426-40b6-98b3-8635f66deb53.png)

(Yes, I rebooted many times, which is why the socket closes).  Also, If no traffic is sent for 60 seconds, the connection is closed via the load balancer.  This is configurable from the load balancer side. Since the .Net side is not closing the socket, eventually the server crashes with `503.2 Too many connections.`  

Any assistance in a quick resolution would be awesome otherwise, I have to downgrade back to V3.

**Comment by Shane32** [comment]
Can you provide the configuration and setup you are using as it relates to the IIS pipeline?  Probably everything in the `public void Configure(IApplicationBuilder app, IWebHostEnvironment env)` method would be useful.  Also, which exact versions of GraphQL.NET and GraphQL Server you have installed.  I'd be happy to work with you to get this solved -- thanks!!

**Comment by Shane32** [comment]
Also wondering if you are running on a Windows or a Linux server, and if you use websockets with subscriptions.

**Comment by STRATZ-Ken** [comment]
@Shane32 Thanks, is there anyway we can meet in real time? Discord? Google Hangout?

Windows OS, Windows 2012 R2.

Versions :

```
<ItemGroup>
    <PackageReference Include="AspNet.Security.OpenId.Steam" Version="5.0.0" />
    <PackageReference Include="graphiql" Version="2.0.0" />
    <PackageReference Include="GraphQL" Version="4.6.0" />
    <PackageReference Include="GraphQL-Parser" Version="7.2.0" />
    <PackageReference Include="GraphQL.DataLoader" Version="4.6.0" />
    <PackageReference Include="GraphQL.NewtonsoftJson" Version="4.6.0" />
    <PackageReference Include="GraphQL.Server.Core" Version="5.0.2" />
    <PackageReference Include="GraphQL.Server.Transports.AspNetCore" Version="5.0.2" />
    <PackageReference Include="GraphQL.Server.Transports.AspNetCore.SystemTextJson" Version="5.0.2" />
    <PackageReference Include="GraphQL.Server.Transports.Subscriptions.Abstractions" Version="5.0.2" />
    <PackageReference Include="GraphQL.Server.Transports.Subscriptions.WebSockets" Version="5.0.2" />
    <PackageReference Include="GraphQL.SystemReactive" Version="4.6.0" />
    <PackageReference Include="LazyCache.AspNetCore" Version="2.4.0" />
    <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="5.0.10" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="5.0.10" />
    <PackageReference Include="Microsoft.OpenApi" Version="1.2.3" />
    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="5.0.2" />
    <PackageReference Include="NLog" Version="4.7.11" />
    <PackageReference Include="NLog.Web.AspNetCore" Version="4.14.0" />
    <PackageReference Include="Sentry.AspNetCore" Version="3.9.3" />
    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.2.2" />
    <PackageReference Include="System.IO.Pipelines" Version="5.0.1" />
    <PackageReference Include="System.Threading.Tasks.Dataflow" Version="5.0.0" />
    <PackageReference Include="tusdotnet" Version="2.4.0" />
  </ItemGroup>
```

```
app.UseWebSockets();
            app.UseGraphQLWebSockets<DotaSchema>();
            app.UseAuthentication();

            if (env.IsDevelopment())
            {
                _logger.Debug("Configure: Evnironment is development");
                app.UseDeveloperExceptionPage();
                app.UseCors("debug");
            }

            Directory.CreateDirectory(@"uploads");
            app.UseTus(httpContext => new DefaultTusConfiguration
            {
                // TUS CODE HERE
            });

            app.UseSwagger(c =>
            {
                c.PreSerializeFilters.Add((swagger, httpReq) =>
                {
                    swagger.Servers = new List<OpenApiServer> { new OpenApiServer { Url = Configuration["TokenIssuer"] } };
                });

            });
            // Enable middleware to serve swagger-ui (HTML, JS, CSS, etc.), specifying the Swagger JSON endpoint.
            app.UseSwaggerUI(c =>
            {
                c.SwaggerEndpoint("/swagger/v1/swagger.json", "Dota API");
                c.RoutePrefix = "";
            });

            //app.UseHttpsRedirection();

            app.UseRouting();

            app.UseAuthorization();

            // This has to go before the GraphQL setup, or GraphQL will not be compressed
            app.UseResponseCompression();

            app.UseGraphQL<DotaSchema>();
            app.UseGraphiQl("/graphiql", "/graphql");

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
            });
```

**Comment by STRATZ-Ken** [comment]
Also worth noting that the system hits a KONG server (https://docs.konghq.com/hub/kong-inc/rate-limiting/) first, then hits a HAProxy load balancer then hits GraphQL (.Net Core 5.0) servers.  

Lastly, worth mentioning, you dont have to reboot the Application Pool to fix the issue, you just need to "restart"..

![image](https://user-images.githubusercontent.com/24981405/134666159-e391e9ce-e7c9-4432-9783-5976e07b714b.png)

**Comment by STRATZ-Ken** [comment]
I also just removed WebSockets from my application and deployed to production:

![image](https://user-images.githubusercontent.com/24981405/134670640-a8576b5c-8627-48f5-816f-9b2e8c37c738.png)


As you can see about 57~ users right now.  Just to double double confirm it has to do with the WebSockets/Subscription model.

**Comment by Shane32** [comment]
@STRATZ-Ken add me to discord at “Lt Cmdr Data#8576” and look for me in about 35 minutes from now

**Comment by Shane32** [comment]
Notes so far: an excessive number of inbound HTTPS connections with status of `CLOSE_WAIT` are observed via `netstat`.  There should not be _any_ in that state typically, as this indicates that the TCP connection has been closed remotely, and the system is waiting for the local service/application to acknowledge that the socket has been remotely closed.  This is likely due to some of the .NET Websocket code failing to close the underlying socket.

See:

https://stackoverflow.com/questions/64385742/system-net-websockets-leaving-sockets-in-close-wait-after-disconnect

**Comment by Shane32** [comment]
Ref: 
- #561

**Comment by STRATZ-Ken** [comment]
I was able to mitigate the damage but writing a script to reboot the Application Pool (Recycle) every 15 minutes to release the connections. Far from perfect but at least working.

**Comment by wpoPR** [comment]
Hey Guys, any update so far? looks like we hit the same issue here!
we were holding 1.5K opened connections, then when we restart our pods on the cluster the number decrease!

**Comment by STRATZ-Ken** [comment]
Nothing found as of yet.  @wpoPR Do you run behind a load balancer? HAProxy of any sort?

**Comment by wpoPR** [comment]
we use Nginx. we restart our pods manually and the connections just decrease 90%, we are still monitoring and considering restarting every single day with some cronJob

**Comment by Shane32** [comment]
> Hey Guys, any update so far? looks like we hit the same issue here!

We need some help diagnosing this issue.  This is what I would do, if I had the time:

1. Open or create a sample 'chat' application -- any very basic sample that uses GraphQL subscriptions
2. Run `netstat -a -n` and verify that there are none listed with status of `CLOSE_WAIT`
3. Run the server and then have a few clients connect and disconnect
4. Run `netstat -a -n` and see if there are any connections with status of `CLOSE_WAIT`
5. If none appear, try across a network instead of localhost

Hopefully the problem can be reproduced locally with the above sample setup.  Assuming this is so, then I'd experiment with solutions, such as:

1. Ensure that the `WebSocket` instance is being closed - either with breakpoints or by adding `Debug.WriteLine` throughout
2. Try explicitly closing the underlying `Socket`, if possible, that's behind the `WebSocket` instance.
3. Try some of the solutions listed on the stackoverflow article listed above.

I'd really like to get this fixed, then release an updated version of the server project.  I just have not had the time to work on it.

If either of you have time to work on this, please continue to post progress notes here; what you've tried or not, GitHub links for sample projects, whatever, and we can all collaborate to get this fixed.  Thanks!!!!

**Comment by Shane32** [comment]
Once a solution is found, we should ensure that it works correctly whether the connection is gracefully terminated from the server side or the client side, or if the client connection is unexpectedly lost (i.e. no TCP FIN packet received by the server).

**Comment by STRATZ-Ken** [comment]
Sorry, I have not had time to look into this currently.  Our largest event of the year is coming up and was hopeful @wpoPR Had a few CPU cycles to take a look.

**Comment by Shane32** [comment]
Last weekend I tried some limited testing along the lines that I outlined above, and was unable to reproduce the problem on my dev machine.  I didn't try establishing a connection across a physical network, or breaking a live connection at the TCP layer.

Of note, it is interesting that both of your cases involve use of a load balancer.  I don't know how that could affect this issue, but it is noteworthy.

**Comment by wpoPR** [comment]
just an update from my side that after 3 days we hit the limit of connections, manually I restarted our pods and we reduce ~70% the number. I'm trying here some approaches on the database trying to set to identify idle connections and kill them! I let you guys know.

**Comment by Shane32** [comment]
See possible related issue:
- #667

**Comment by Shane32** [comment]
@STRATZ-Ken if you want to try this beta build here, it may solve your issue:

https://github.com/graphql-dotnet/server/packages/493603?version=5.1.0-preview-126

**Comment by sungam3r** [comment]
Should be 5.1.2-preview

**Comment by Shane32** [comment]
Thanks for testing @STRATZ-Ken .  I didn't think it would -- it seemed like it was solving a different memory leak than the one you are experiencing.  But wished it would have of course!

**Comment by STRATZ-Ken** [comment]
Spent a lot of the morning rebuilding the entire project locally from scratch.  

- Created basic Server API Implementation with the latest code from this GIT.
- Create a basic application in C# that connect via subscription. This application has the ability of multiple threads for stress testing.
- Installed HAProxy locally and setup a forward to my IIS Express from the client application.

I can confirm its only HAProxy causing the issue with the Aborted sockets.  When a socket gets aborted, the `WebSocketReaderPipeline.cs` Line 47 `if` statement gets called twice.  Obviously since its aborted, it does not enter the statement.  I did not see any thrown exceptions anywhere.  

You can replicate the issue by setting in haproxy main config file the `timeout client` and `timeout server` to a low value, and do not send a `OnNext` command for the more than allotted time in the Subscription service.   

I was able to stabilize my API a bit, by increasing the timeout from 30 to 180seconds as a temporary fix on production.  However, if a socket is created and no data is sent for 180 seconds, I would have 1 instance of the leak occur.  

I wanted to try to use Performance monitor to open and see the Web Sockets Active, however, in IIS Express it seems its not possible.  I did not add in any database connections in this test.

If anyone wants a copy of the local example, I can provide that.  

I hope someone else has some answers. Thanks as always!

**Comment by Shane32** [comment]
Do you have a stack trace of how it hit `WebSocketReaderPipeline.cs:47` ?

**Comment by STRATZ-Ken** [comment]
![image](https://user-images.githubusercontent.com/24981405/146818945-780c694e-123b-484b-b7ef-b81580bca402.png)

![image](https://user-images.githubusercontent.com/24981405/146818981-5f5c8bb4-9acd-42cb-b259-c77d60ee1816.png)

**Comment by Shane32** [comment]
What I see the problem is that nowhere is there a call to `WebSocket.Dispose`.  `CloseAsync` and `CloseOutputAsync` are not a replacement for `Dispose`, and `Dispose` is not called if the socket state is "closed", "close sent" or "aborted".

Perhaps in that method a condition could be added "if socket state == aborted, then socket.Dispose" but that may (or may not) cause problems within WebSocketWriterPipeline

**Comment by Shane32** [comment]
See new PR:
- #671

**Comment by STRATZ-Ken** [comment]
@Shane32 Looks like the build doesnt work at all. Sockets don't seem to initially connect.

**Comment by sungam3r** [comment]
> What I see the problem is that nowhere is there a call to WebSocket.Dispose. CloseAsync and CloseOutputAsync are not a replacement for Dispose, and Dispose is not called if the socket state is "closed", "close sent" or "aborted".

Right. Actually `WebSocketTransport` owns the socket, but it doesn't know now when to dispose it. `WebSocketReaderPipeline.Complete(WebSocketCloseStatus closeStatus, string statusDescription)` is a good candidate to dispose socket. In fact even more - to dispose a whole connection object.

**Comment by sungam3r** [comment]
Connection object now is created at the end of `GraphQLWebSocketsMiddleware.InvokeAsync` and abandoned.

**Comment by sungam3r** [comment]
And I think that connection is even garbage-collected after `Connect()` rather soon.

**Comment by Shane32** [comment]
From the notes and code, I had thought that the async `Connect` method would wait until the pipelines disconnected, which is why it ended with `CloseAsync`.  So a finally block to dispose of the websocket should be fine.  That must not be the case.

I really need to run an implementation to test these code changes better.  I guess for now I'll have to revert the PR.

**Comment by STRATZ-Ken** [comment]
@Shane32 I DM'ed you in Discord a full working client/server build. Just adjust the IP to your local computer.

**Comment by Shane32** [comment]
@STRATZ-Ken How is the latest version 5.2.0 working for you?

**Comment by STRATZ-Ken** [comment]
@Shane32 Sorry for the delay'ed response.  I didn't want to approve or disapprove anything without a 48 hour test period.  I think my issue is fully resolved with 5.2.0.  I have no idea what the cause was, but I know the issue was created in the 5.1.X build.  I was on 3.X.X before and never had an issue.  But when moving to 5.1.X the issue was created. 

I guess we will never know.
